pkg/dataobj/internal/dataset/reader_basic.go

```diff
@@ -257,11 +258,25 @@ func reuseRowsBuffer(dst []Value, src []Row, columnIndex int) []Value {
 		if value.IsNil() {
 			continue
 		}
+
+		// Only appending the values to dst without swapping could result in sublte bugs:
+		// Assume the column is sparsely polulated with non-null values
+		// 1. A non-null value from src[500] could be copied to dst[5].
+		// 2. When the reader writes back dst[5] to src[5], we now have the same value at both src[5] and src[500]
+		//    This also assumes fewer than requested values are read.
+		// 3. This creates memory aliasing - changes to one position would affect the other
+		//
+		// Swapping ensures that the values being reused are in the same position in src and dst.
+		// This way when we write back to src, we do not risk referring to the same location twice.
+		if i != filledLength {
+			src[filledLength].Values[columnIndex], src[i].Values[columnIndex] =
+				value, src[filledLength].Values[columnIndex]
+		}
+
```

Comment:
This is very nice catch :) 

I thought about it a bit and I think my main concern is how this reorders rows from the `src`. Since this is on the `fill` code path, I'm having a hard time reasoning about whether reordering rows would cause any issues. 

As an alternative, what if `reuseRowsBuffer` became `moveRowsBuffer`, so that the memory moves to `dst` for the given column, zeroing it out in the src? Then we move it back into src as usual afterwards. 

This would probably require a small change to `basicReader.fill` so that we always move _all_ memory back to `src`, even if it's beyond the count we read. That way any allocated memory doesn't get dropped just by a short read.

What do you think?

Reply:
Oh wait, I see that it's not changing rows, just the values in them. Hm, let me think on that some more, that might change whether I'm concerned